src/Backend.idr

```diff
@@ -0,0 +1,102 @@
+module Backend
+
+import Data.Vect
+
+import Types
+import Typedefs
+
+import Data.Vect
+import Text.PrettyPrint.WL
+
+%default total
+%access public export
+
+||| A parametric type declaration (no definition, only name and parameters).
+record Decl where
+  constructor MkDecl
+
+  ||| The name of the type.
+  name   : Name
+
+  ||| The names of the type's parameters.
+  params : Vect n Name -- TODO is number of parameters enough?
+
+||| A variable environment. Left=free, Right=bound.
+Env : Nat -> Type
+Env k = Vect k (Either String Decl)
+
+||| Vertically concatenate a list of documents with two newlines (i.e. one empty line) as separator.
+vsep2 : List Doc -> Doc
+vsep2 = vsep . punctuate line
+
+||| Get the free names from the environment.
+getFreeVars : (e : Env n) -> Vect (fst (Vect.filter Either.isLeft e)) String
+getFreeVars e with (filter isLeft e)
+  | (p ** v) = map (either id (const "")) v
+
+||| Get a list of the de Brujin indices that are actually used in a `TDef`.
+getUsedIndices : TDef n -> List (Fin n)
+getUsedIndices T0         = []
+getUsedIndices T1         = []
+getUsedIndices (TSum xs)  = assert_total $ concatMap getUsedIndices xs
+getUsedIndices (TProd xs) = assert_total $ concatMap getUsedIndices xs
+getUsedIndices (TVar i)   = [i]
+getUsedIndices (TMu _ xs) = assert_total $ concatMap ((concatMap weedOutZero) . getUsedIndices . snd) xs
+  where weedOutZero : Fin (S n) -> List (Fin n)
+        weedOutZero FZ     = []
+        weedOutZero (FS i) = [i]
+getUsedIndices (TName _ t) = getUsedIndices t
+
+||| Filter out the entries in an `Env` that is referred to by a `TDef`.
+getUsedVars : Env n -> (td: TDef n) -> Env (length (getUsedIndices td))
+getUsedVars e td = map (flip index e) (fromList $ getUsedIndices td)
+
+||| Interface for codegens. `lang` is a type representing (the syntactic structure of)
+||| a type declaration in the target language.
+interface Backend lang where
+  ||| Given a TDef and a matching environment, generate a list of type definitions
+  ||| in the target language.
+  generateTyDefs : Env n -> TDef n -> List lang
+
+  ||| Given a type definition in the target language, generate its code.
+  generateCode : lang -> Doc
+
+  ||| Generate a new environment with n free names.
+  freshEnv : (n: Nat) -> Env n
+
+||| Generate code in a specific language for a type definition and all its dependencies.
+||| Needs to be called with the implicit parameter `lang`, as such: `generate {lang=Haskell} td`.
```

Comment:
I actually tried to do this, but didn't succeed. I tried again now, with the following code:

```idris
generate : Backend lang => {n: Nat} -> (lang: Type) -> TDef n -> Doc
generate {n} lang td = vsep2 . map (generateCode) . generateTyDefs {lang=lang} (freshEnv {lang=lang} n) $ td
```

Which gives me the error
```
Backend.idr:70:24-108:
   |
70 | generate {n} lang td = vsep2 . map (generateCode) . generateTyDefs {lang=lang} (freshEnv {lang=lang} n) $ td
   |                        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
When checking right hand side of generate with expected type
        Doc

When checking an application of function Prelude.Basics..:
        Can't find implementation for Backend lang
```

I understand this as that it isn't inferred that `Backend lang` and `(lang: Type)` refer to the same type. If you know how to solve this I'm all ears, because I agree that the current situation is ugly.

Reply:
Good point, though I think addressing this in a separate PR might be good enough. Perhaps we could just make an issue out of this?

Reply:
@kjekac this is because you shadow over lang with you later implicit. If you write it as 
```
generate : (lang: Type) -> Backend lang => {n: Nat} -> TDef n -> Doc
generate lang {n} td = vsep2 . map (generateCode) . generateTyDefs {lang} (freshEnv {lang} n) $ td
```
it seems to work.

Reply:
Also a minor nitpick of mine is that you can always shorten `{foo=foo}` to just `{foo}` :)

Reply:
Thanks Alex! Pushed an update addressing both this and your nitpicks. :)